fluid/ocr_ctc/train.py

```diff
@@ -0,0 +1,187 @@
+    flattened_data = flattened_data.reshape([len(flattened_data), 1])
+    res = core.LoDTensor()
+    res.set(flattened_data, place)
+    res.set_lod([lod])
+    return res
+
+def _get_feeder_data(data, place):
+    pixel_tensor = core.LoDTensor()
+    pixel_data = np.concatenate(
+        map(lambda x: x[0][np.newaxis, :], data), axis=0).astype("float32")
+    pixel_tensor.set(pixel_data, place)
+    label_tensor = _to_lodtensor(map(lambda x: x[1], data), place)
+    return {"pixel": pixel_tensor, "label": label_tensor}
+
+def _ocr_conv(input, num, with_bn, param_attrs):
+    assert (num % 4 == 0)
+
+    def _conv_block(input, filter_size, group_size, with_bn):
+        return fluid.nets.img_conv_group(
+            input=input,
+            conv_num_filter=[filter_size] * group_size,
+            pool_size=2,
+            pool_stride=2,
+            conv_padding=1,
+            conv_filter_size=3,
+            conv_act='relu',
+            conv_with_batchnorm=with_bn,
+            pool_type='max',
+            param_attr=param_attrs)
+
+    conv1 = _conv_block(input, 16, (num / 4), with_bn)
+    conv2 = _conv_block(conv1, 32, (num / 4), with_bn)
+    conv3 = _conv_block(conv2, 64, (num / 4), with_bn)
+    conv4 = _conv_block(conv3, 128, (num / 4), with_bn)
+    return conv4
+
+
+def _ocr_ctc_net(images, num_classes, param_attrs):
+    conv_features = _ocr_conv(images, 8, True, param_attrs)
+    sliced_feature = fluid.layers.im2sequence(
+        input=conv_features, stride=[1, 1], filter_size=[1, 3])
+    gru_forward = fluid.layers.dynamic_gru(
+        input=sliced_feature, size=128, param_attr=param_attrs)
+    gru_backward = fluid.layers.dynamic_gru(
+        input=sliced_feature, size=128, is_reverse=True, param_attr=param_attrs)
+
+    fc_out = fluid.layers.fc(input=[gru_forward, gru_backward],
+                             size=num_classes + 1,
+                             param_attr=param_attrs)
+    return fc_out
+
+
+
+def train(l2=0.0005,
+          min_clip=-10,
+          max_clip=10,
+          data_reader=dummy_reader,
+          learning_rate=1.0e-3,
+          momentum=0.9,
+          batch_size=16,
+          pass_num=2,
+          device=0):
+    """OCR CTC training"""
+    num_classes = data_reader.num_classes()
+    # define network
+    param_attrs = fluid.ParamAttr(
+        regularizer=fluid.regularizer.L2Decay(l2 * batch_size),
+        gradient_clip=fluid.clip.GradientClipByValue(max_clip, min_clip))
+    data_shape = data_reader.data_shape()
+    images = fluid.layers.data(name='pixel', shape=data_shape, dtype='float32')
+    label = fluid.layers.data(
+        name='label', shape=[1], dtype='int32', lod_level=1)
+    fc_out = _ocr_ctc_net(images, num_classes, param_attrs)
+
+    # define cost and optimizer
+    cost = fluid.layers.warpctc(
+        input=fc_out,
+        label=label,
+        size=num_classes + 1,
+        blank=num_classes,
+        norm_by_times=True)
```

Comment:
norm_by_times=True -> norm_by_times=False?

Reply:
`norm_by_times ` means whether to divide gradients by sequence length.
With `mean` op, gradients were divided by batch_size. 
If we want to avoid the effect of `mean` op, it's more reasonable to remove `mean_grad` op but not make  `norm_by_times=False`.

Reply:
Actually, in current code, the target to be minimized by optimizer is `cost` but not `avg_cost`.
```
# define cost and optimizer
113     cost = fluid.layers.warpctc(
114         input=fc_out,
115         label=label,
116         size=num_classes + 1,
117         blank=num_classes,
118         norm_by_times=True)
119     avg_cost = fluid.layers.mean(x=cost)
120     optimizer = fluid.optimizer.Momentum(
121         learning_rate=args.learning_rate, momentum=args.momentum)
122     opts = optimizer.minimize(cost)
```